File: python/parmana/api/verification_api.py
"""
Parmana Verification API.

Verify Execution Trust Records.
"""

from __future__ import annotations

from parmana.config.transport import Transport
from parmana.models.verification import Verification


class VerificationApi:
    """
    Verification API.

    Responsibilities
    ----------------
    - Verify Execution Trust Records (fresh verification)
    - Retrieve the latest cached Verification

    This API does NOT:
    - execute Business Transactions
    - replay executions
    - generate receipts
    - validate policies
    """

    def __init__(
        self,
        transport: Transport,
    ) -> None:
        self._transport = transport

    def verify(
        self,
        business_transaction_id: str,
    ) -> Verification:
        """
        Run a fresh verification of an Execution Trust Record.

        Maps to POST /verify. Each call re-verifies the record and
        appends a new Verification to its history.

        Parameters
        ----------
        business_transaction_id:
            Business Transaction identifier.

        Returns
        -------
        Verification.
        """

        return self._transport.send(
            method="POST",
            path="/verify",
            body={
                "businessTransactionId": business_transaction_id,
            },
            response_model=Verification,
        )

    def get_latest(
        self,
        business_transaction_id: str,
    ) -> Verification:
        """
        Retrieve the most recent Verification without re-verifying.

        Maps to GET /verification/:id, distinct from `verify()` (POST
        /verify), which performs a fresh verification.

        Parameters
        ----------
        business_transaction_id:
            Business Transaction identifier.

        Returns
        -------
        Verification.
        """

        return self._transport.send(
            method="GET",
            path=f"/verification/{business_transaction_id}",
            response_model=Verification,
        )

File: typescript/src/client/VerificationApi.ts
/**
 * Parmana SDK
 *
 * Verification API.
 */

import type {
  Verification,
} from "../models/index.js";

import type {
  Transport,
} from "../config/Transport.js";

export class VerificationApi {
  constructor(
    private readonly transport: Transport,
  ) {}

  /**
   * Runs a fresh verification of an Execution Trust Record. Maps to
   * POST /verify. Each call re-verifies the record and appends a new
   * Verification to its history, distinct from getLatest() (GET
   * /verification/:id), which reads the most recent one without
   * re-verifying.
   */
  public async verify(
    businessTransactionId: string,
  ): Promise<Verification> {
    const response =
      await this.transport.send<Verification>({
        method: "POST",
        path: "/verify",
        body: {
          businessTransactionId,
        },
      });

    return response.body;
  }

  /**
   * Returns the latest Verification for a
   * Business Transaction, without performing a
   * fresh verification. Maps to GET
   * /verification/:id.
   */
  public async getLatest(
    businessTransactionId: string,
  ): Promise<Verification> {
    const response =
      await this.transport.send<Verification>({
        path: `/verification/${businessTransactionId}`,
        method: "GET",
      });

    return response.body;
  }
}